In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt
import seaborn as sns
import datetime
import time
import glob
%matplotlib inline
from sklearn import preprocessing
import tables
import h5py
from sqlalchemy import create_engine

In [2]:
flights = pd.read_csv('Big_Data_PrunedV2.csv')
flights.head()

C:\Users\andre_000\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2717: DtypeWarning: Columns (9,10,11,12,14,15,19,20,21,22,24,25,29,30,31) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


MONTH  DOW AIRLINE ORIGIN_AIRPORT DESTINATION_AIRPORT  DISTANCE  \
0      1    4      AS            ANC                 SEA      1448   
1      1    4      AS            ANC                 SEA      1448   
2      1    4      DL            ANC                 SEA      1448   
3      1    4      AS            GEG                 SEA       224   
4      1    4      AA            LAX                 PBI      2330   

   ARRIVAL_DELAY  DIVERTED  CANCELLED dept_temp     ...      arr_vsby  \
0          -22.0         0          0      33.8     ...          10.0   
1          -14.0         0          0      33.8     ...          10.0   
2          -24.0         0          0      33.8     ...          10.0   
3            0.0         0          0      8.96     ...          10.0   
4           -9.0         0          0     46.04     ...         10.00   

  arr_gust arr_sky  arr_weather CertificateDate    Manu      Model DAYS  \
0        M     CLR            M            2012  BOEING  737-990ER    0   
1        M     CLR            M               0       0          0    0   
2        M     CLR            M            2001  AIRBUS   A321-211    0   
3        M     CLR            M               0       0          0    0   
4        M     FEW            M            2009  BOEING    737-890    0   

  DEPARTURE_HOUR ARRIVAL_HOUR  
0            9.1         13.5  
1            9.8         14.2  
2            9.8         14.2  
3           13.0         14.2  
4            8.2         15.8  

[5 rows x 35 columns]

In [16]:
del flights["dept_weather"]
del flights["arr_weather"]
del flights["dept_gust"]
del flights["arr_gust"]

In [19]:
cols = list(flights)
cols

['MONTH',
 'DOW',
 'AIRLINE',
 'ORIGIN_AIRPORT',
 'DESTINATION_AIRPORT',
 'DISTANCE',
 'ARRIVAL_DELAY',
 'DIVERTED',
 'CANCELLED',
 'dept_temp',
 'dept_dew',
 'dept_hum',
 'dept_windspd',
 'dept_precip',
 'dept_alti',
 'dept_vsby',
 'dept_sky',
 'arr_temp',
 'arr_dew',
 'arr_hum',
 'arr_windspd',
 'arr_precip',
 'arr_alti',
 'arr_vsby',
 'arr_sky',
 'CertificateDate',
 'Manu',
 'Model',
 'DAYS',
 'DEPARTURE_HOUR',
 'ARRIVAL_HOUR']

In [18]:
colsToSee = cols[9:-6]
colsToSee

['dept_temp',
 'dept_dew',
 'dept_hum',
 'dept_windspd',
 'dept_precip',
 'dept_alti',
 'dept_vsby',
 'dept_sky',
 'arr_temp',
 'arr_dew',
 'arr_hum',
 'arr_windspd',
 'arr_precip',
 'arr_alti',
 'arr_vsby',
 'arr_sky']

In [20]:
numDrop = set()
for i in colsToSee:
    for i, j in enumerate(flights[i]):
        if j == 'M' or (not isinstance(j, str) and np.isnan(float(j))):
            numDrop.add(i)
print(numDrop)
        

{130259, 4587530, 1572879, 2490395, 4194333, 2883618, 5242926, 1703986, 5636153, 4063296, 4194368, 1572957, 655456, 3670115, 3801195, 1572979, 2621556, 3932280, 3407995, 1310848, 2097290, 4063373, 3932302, 655507, 3932312, 1441976, 5636291, 2621646, 4849872, 4849873, 4849874, 1310931, 4849875, 4849876, 1310944, 1310945, 3277039, 3408149, 3408150, 2621725, 3408174, 4325702, 393542, 3539281, 5636433, 5636448, 4850020, 2228588, 3670383, 3670384, 1704311, 1573239, 1573240, 1573241, 5636478, 131458, 131459, 5374362, 3277211, 2490781, 2752934, 4194734, 1311156, 4325813, 3801529, 1442236, 4850114, 4063686, 4194764, 4456915, 2884064, 2884070, 262640, 4194825, 2359820, 3408399, 3801615, 548, 5767718, 4850216, 554, 5636653, 2359859, 3801664, 4850240, 4850241, 4850242, 4850243, 4850244, 3408463, 4850256, 655961, 655962, 655963, 3932764, 655964, 655965, 4850278, 4194920, 1966703, 1966704, 1966705, 3539575, 4194974, 1180325, 1573542, 1966768, 262833, 1966770, 3801804, 1180398, 3015406, 262896, 1311

In [21]:
len(numDrop)

23412

In [22]:
flights.drop(flights.index[list(numDrop)], inplace=True)

In [25]:
flights.head()

MONTH  DOW AIRLINE ORIGIN_AIRPORT DESTINATION_AIRPORT  DISTANCE  \
0      1    4      AS            ANC                 SEA      1448   
1      1    4      AS            ANC                 SEA      1448   
2      1    4      DL            ANC                 SEA      1448   
3      1    4      AS            GEG                 SEA       224   
4      1    4      AA            LAX                 PBI      2330   

   ARRIVAL_DELAY  DIVERTED  CANCELLED dept_temp     ...      arr_precip  \
0          -22.0         0          0      33.8     ...               0   
1          -14.0         0          0      33.8     ...               0   
2          -24.0         0          0      33.8     ...               0   
3            0.0         0          0      8.96     ...               0   
4           -9.0         0          0     46.04     ...               0   

  arr_alti arr_vsby  arr_sky CertificateDate    Manu      Model DAYS  \
0    30.32     10.0      CLR            2012  BOEING  737-990ER    0   
1    30.32     10.0      CLR               0       0          0    0   
2    30.32     10.0      CLR            2001  AIRBUS   A321-211    0   
3    30.32     10.0      CLR               0       0          0    0   
4    30.21    10.00      FEW            2009  BOEING    737-890    0   

  DEPARTURE_HOUR ARRIVAL_HOUR  
0            9.1         13.5  
1            9.8         14.2  
2            9.8         14.2  
3           13.0         14.2  
4            8.2         15.8  

[5 rows x 31 columns]

In [28]:
len(flights)

5795667

In [29]:
flights.to_csv("BigData_Pruned_drop.csv", index = False)

In [32]:
additionalDrop = []
for a,i in enumerate(flights["Manu"]):
    if i == "0":
        additionalDrop.append(a)

In [33]:
flights.drop(flights.index[additionalDrop], inplace=True)

In [35]:
flights.to_csv("BigData_Pruned_drop_Manu.csv", index = False)

In [34]:
sum(1 for i in flights["Manu"] if i == '0')

0

In [5]:
sum(1 for i in flights["Manu"] if i == '0')

805464

In [3]:
delays = []
for i in flights["ARRIVAL_DELAY"]:
    if np.isnan(i):
        delays.append(180)
    else:
        delays.append(int(i))

In [4]:
flights["ARRIVAL_DELAY"] = delays

In [8]:
windSpeeds = []
for i in flights["dept_windspd"]:
    if i == "M":
        windSpeeds.append(7)
    else:
        windSpeeds.append(int(float(i)))

In [9]:
flights["dept_windspd"] = windSpeeds

In [13]:
windSpeeds = []
for i in flights["arr_windspd"]:
    if i == "M" or np.isnan(float(i)):
        windSpeeds.append(7)
    else:
        windSpeeds.append(int(float(i)))

In [14]:
flights["arr_windspd"] = windSpeeds

In [22]:
alt = []
for i in flights["dept_alti"]:
    if i == "M":
        alt.append(29.57)
    else:
        alt.append(float(i))

In [23]:
flights["dept_alti"] = alt

In [ ]:
alt = []
for i in flights["arr_alti"]:
    if i == "M":
        alt.append(30.03)
    else:
        alt.append(float(i))

In [ ]:
flights["arr_alti"] = alt

In [29]:
lst = []
for i in flights["dept_temp"]:
    if i == "M":
        lst.append("M")
    else:
        lst.append(float(i))

In [30]:
sum(float(i) for i in flights["dept_temp"] if i != 'M') / sum(1 for i in flights["dept_temp"] if i != 'M')

56.41847169152798

Dept_Windspeed: 7.61224751836367 -> 7
Arr_Windspeed: 7.680535201372598 -> 7
Dept_alt -> 29.57273861595442 -> 29.57
Arr_alt -> 30.03397670547148 -> 30.03
dept_temp -> 56.41847169152798 -> 56.42


In [6]:
orig_airports = flights['ORIGIN_AIRPORT'].tolist()
dest_airports = flights['DESTINATION_AIRPORT'].tolist()

In [8]:
le_airports = preprocessing.LabelEncoder()
le_airports.fit(orig_airports)
#print(le_airports.classes_)
enc_orig_airports = le_airports.transform(orig_airports)
enc_dest_airports = le_airports.transform(dest_airports) 

enc_dest_airports

['ABE' 'ABI' 'ABQ' 'ABR' 'ABY' 'ACK' 'ACT' 'ACV' 'ACY' 'ADK' 'ADQ' 'AEX'
 'AGS' 'AKN' 'ALB' 'ALO' 'AMA' 'ANC' 'APN' 'ASE' 'ATL' 'ATW' 'AUS' 'AVL'
 'AVP' 'AZO' 'BDL' 'BET' 'BFL' 'BGM' 'BGR' 'BHM' 'BIL' 'BIS' 'BJI' 'BLI'
 'BMI' 'BNA' 'BOI' 'BOS' 'BPT' 'BQK' 'BQN' 'BRD' 'BRO' 'BRW' 'BTM' 'BTR'
 'BTV' 'BUF' 'BUR' 'BWI' 'BZN' 'CAE' 'CAK' 'CDC' 'CDV' 'CEC' 'CHA' 'CHO'
 'CHS' 'CID' 'CIU' 'CLD' 'CLE' 'CLL' 'CLT' 'CMH' 'CMI' 'CMX' 'CNY' 'COD'
 'COS' 'COU' 'CPR' 'CRP' 'CRW' 'CSG' 'CVG' 'CWA' 'DAB' 'DAL' 'DAY' 'DBQ'
 'DCA' 'DEN' 'DFW' 'DHN' 'DIK' 'DLG' 'DLH' 'DRO' 'DSM' 'DTW' 'DVL' 'EAU'
 'ECP' 'EGE' 'EKO' 'ELM' 'ELP' 'ERI' 'ESC' 'EUG' 'EVV' 'EWN' 'EWR' 'EYW'
 'FAI' 'FAR' 'FAT' 'FAY' 'FCA' 'FLG' 'FLL' 'FNT' 'FSD' 'FSM' 'FWA' 'GCC'
 'GCK' 'GEG' 'GFK' 'GGG' 'GJT' 'GNV' 'GPT' 'GRB' 'GRI' 'GRK' 'GRR' 'GSO'
 'GSP' 'GST' 'GTF' 'GTR' 'GUC' 'GUM' 'HDN' 'HIB' 'HLN' 'HNL' 'HOB' 'HOU'
 'HPN' 'HRL' 'HSV' 'HYA' 'HYS' 'IAD' 'IAG' 'IAH' 'ICT' 'IDA' 'ILG' 'ILM'
 'IMT' 'IND' 'INL' 'ISN' 'ISP' 'ITH' 'ITO' 'JAC' 'J

array([277, 277, 277, ..., 303, 203, 285], dtype=int64)

In [10]:
flights['ORIGIN_AIRPORT'] = enc_orig_airports
flights['DESTINATION_AIRPORT'] = enc_dest_airports
flights.head()

MONTH  DOW AIRLINE  ORIGIN_AIRPORT  DESTINATION_AIRPORT  DISTANCE  \
0      1    4      AS              17                  277      1448   
1      1    4      AS              17                  277      1448   
2      1    4      DL              17                  277      1448   
3      1    4      AS             121                  277       224   
4      1    4      AA             176                  235      2330   

   ARRIVAL_DELAY  DIVERTED  CANCELLED dept_temp     ...      arr_vsby  \
0          -22.0         0          0      33.8     ...          10.0   
1          -14.0         0          0      33.8     ...          10.0   
2          -24.0         0          0      33.8     ...          10.0   
3            0.0         0          0      8.96     ...          10.0   
4           -9.0         0          0     46.04     ...         10.00   

  arr_gust arr_sky  arr_weather CertificateDate    Manu      Model DAYS  \
0        M     CLR            M      12/18/2012  BOEING  737-990ER    0   
1        M     CLR            M               0       0          0    0   
2        M     CLR            M      04/10/2001  AIRBUS   A321-211    0   
3        M     CLR            M               0       0          0    0   
4        M     FEW            M      05/26/2009  BOEING    737-890    0   

  DEPARTURE_HOUR ARRIVAL_HOUR  
0            9.1         13.5  
1            9.8         14.2  
2            9.8         14.2  
3           13.0         14.2  
4            8.2         15.8  

[5 rows x 35 columns]

In [13]:
manufacts = flights['Manu'].tolist()

le_manu = preprocessing.LabelEncoder()
le_manu.fit(manufacts)
#print(le_airports.classes_)
enc_manufacts = le_manu.transform(manufacts)

flights['Manu'] = enc_manufacts
flights.head()

MONTH  DOW AIRLINE  ORIGIN_AIRPORT  DESTINATION_AIRPORT  DISTANCE  \
0      1    4      AS              17                  277      1448   
1      1    4      AS              17                  277      1448   
2      1    4      DL              17                  277      1448   
3      1    4      AS             121                  277       224   
4      1    4      AA             176                  235      2330   

   ARRIVAL_DELAY  DIVERTED  CANCELLED dept_temp     ...      arr_vsby  \
0          -22.0         0          0      33.8     ...          10.0   
1          -14.0         0          0      33.8     ...          10.0   
2          -24.0         0          0      33.8     ...          10.0   
3            0.0         0          0      8.96     ...          10.0   
4           -9.0         0          0     46.04     ...         10.00   

  arr_gust arr_sky  arr_weather CertificateDate Manu      Model DAYS  \
0        M     CLR            M      12/18/2012    7  737-990ER    0   
1        M     CLR            M               0    0          0    0   
2        M     CLR            M      04/10/2001    2   A321-211    0   
3        M     CLR            M               0    0          0    0   
4        M     FEW            M      05/26/2009    7    737-890    0   

  DEPARTURE_HOUR ARRIVAL_HOUR  
0            9.1         13.5  
1            9.8         14.2  
2            9.8         14.2  
3           13.0         14.2  
4            8.2         15.8  

[5 rows x 35 columns]

In [14]:
models = flights['Model'].tolist()

le_model = preprocessing.LabelEncoder()
le_model.fit(models)
#print(le_airports.classes_)
enc_models = le_model.transform(models)

flights['Model'] = enc_models
flights.head()

MONTH  DOW AIRLINE  ORIGIN_AIRPORT  DESTINATION_AIRPORT  DISTANCE  \
0      1    4      AS              17                  277      1448   
1      1    4      AS              17                  277      1448   
2      1    4      DL              17                  277      1448   
3      1    4      AS             121                  277       224   
4      1    4      AA             176                  235      2330   

   ARRIVAL_DELAY  DIVERTED  CANCELLED dept_temp     ...      arr_vsby  \
0          -22.0         0          0      33.8     ...          10.0   
1          -14.0         0          0      33.8     ...          10.0   
2          -24.0         0          0      33.8     ...          10.0   
3            0.0         0          0      8.96     ...          10.0   
4           -9.0         0          0     46.04     ...         10.00   

  arr_gust arr_sky  arr_weather CertificateDate Manu Model DAYS  \
0        M     CLR            M      12/18/2012    7    55    0   
1        M     CLR            M               0    0     0    0   
2        M     CLR            M      04/10/2001    2   105    0   
3        M     CLR            M               0    0     0    0   
4        M     FEW            M      05/26/2009    7    48    0   

  DEPARTURE_HOUR ARRIVAL_HOUR  
0            9.1         13.5  
1            9.8         14.2  
2            9.8         14.2  
3           13.0         14.2  
4            8.2         15.8  

[5 rows x 35 columns]

In [18]:
arr_weather = flights['arr_weather'].tolist()
dept_weather = flights['dept_weather'].tolist()

le_arr_weather = preprocessing.LabelEncoder()
le_arr_weather.fit(arr_weather)
#print(le_airports.classes_)
enc_arr_weather = le_arr_weather.transform(arr_weather)

le_dept_weather = preprocessing.LabelEncoder()
le_dept_weather.fit(dept_weather)
#print(le_airports.classes_)
enc_dept_weather = le_dept_weather.transform(dept_weather)

flights['arr_weather'] = enc_arr_weather
flights['dept_weather'] = enc_dept_weather
flights.head()

MONTH  DOW AIRLINE  ORIGIN_AIRPORT  DESTINATION_AIRPORT  DISTANCE  \
0      1    4      AS              17                  277      1448   
1      1    4      AS              17                  277      1448   
2      1    4      DL              17                  277      1448   
3      1    4      AS             121                  277       224   
4      1    4      AA             176                  235      2330   

   ARRIVAL_DELAY  DIVERTED  CANCELLED dept_temp     ...      arr_vsby  \
0          -22.0         0          0      33.8     ...          10.0   
1          -14.0         0          0      33.8     ...          10.0   
2          -24.0         0          0      33.8     ...          10.0   
3            0.0         0          0      8.96     ...          10.0   
4           -9.0         0          0     46.04     ...         10.00   

  arr_gust arr_sky  arr_weather CertificateDate Manu Model DAYS  \
0        M     CLR          257      12/18/2012    7    55    0   
1        M     CLR          257               0    0     0    0   
2        M     CLR          257      04/10/2001    2   105    0   
3        M     CLR          257               0    0     0    0   
4        M     FEW          257      05/26/2009    7    48    0   

   DEPARTURE_HOUR ARRIVAL_HOUR  
0             9.1         13.5  
1             9.8         14.2  
2             9.8         14.2  
3            13.0         14.2  
4             8.2         15.8  

[5 rows x 35 columns]

In [21]:
arr_sky = flights['arr_sky'].tolist()
dept_sky = flights['dept_sky'].tolist()

le_sky = preprocessing.LabelEncoder()
le_sky.fit(arr_sky)
#print(le_airports.classes_)
enc_arr_sky = le_sky.transform(arr_sky)
enc_dept_sky = le_sky.transform(dept_sky)

flights['dept_sky'] = enc_dept_sky
flights['arr_sky'] = enc_arr_sky
flights.head()

MONTH  DOW AIRLINE  ORIGIN_AIRPORT  DESTINATION_AIRPORT  DISTANCE  \
0      1    4      AS              17                  277      1448   
1      1    4      AS              17                  277      1448   
2      1    4      DL              17                  277      1448   
3      1    4      AS             121                  277       224   
4      1    4      AA             176                  235      2330   

   ARRIVAL_DELAY  DIVERTED  CANCELLED dept_temp     ...      arr_vsby  \
0          -22.0         0          0      33.8     ...          10.0   
1          -14.0         0          0      33.8     ...          10.0   
2          -24.0         0          0      33.8     ...          10.0   
3            0.0         0          0      8.96     ...          10.0   
4           -9.0         0          0     46.04     ...         10.00   

  arr_gust arr_sky  arr_weather CertificateDate Manu Model  DAYS  \
0        M       1          257      12/18/2012    7    55     0   
1        M       1          257               0    0     0     0   
2        M       1          257      04/10/2001    2   105     0   
3        M       1          257               0    0     0     0   
4        M       2          257      05/26/2009    7    48     0   

   DEPARTURE_HOUR ARRIVAL_HOUR  
0             9.1         13.5  
1             9.8         14.2  
2             9.8         14.2  
3            13.0         14.2  
4             8.2         15.8  

[5 rows x 35 columns]

In [24]:
airlines = flights['AIRLINE'].tolist()
le_airline = preprocessing.LabelEncoder()
le_airline.fit(airlines)
#print(le_airports.classes_)
enc_airlines = le_airline.transform(airlines)

flights['AIRLINE'] = enc_airlines
flights.head()

MONTH  DOW  AIRLINE  ORIGIN_AIRPORT  DESTINATION_AIRPORT  DISTANCE  \
0      1    4        1              17                  277      1448   
1      1    4        1              17                  277      1448   
2      1    4        3              17                  277      1448   
3      1    4        1             121                  277       224   
4      1    4        0             176                  235      2330   

   ARRIVAL_DELAY  DIVERTED  CANCELLED dept_temp     ...      arr_vsby  \
0          -22.0         0          0      33.8     ...          10.0   
1          -14.0         0          0      33.8     ...          10.0   
2          -24.0         0          0      33.8     ...          10.0   
3            0.0         0          0      8.96     ...          10.0   
4           -9.0         0          0     46.04     ...         10.00   

  arr_gust arr_sky  arr_weather CertificateDate Manu Model  DAYS  \
0        M       1          257      12/18/2012    7    55     0   
1        M       1          257               0    0     0     0   
2        M       1          257      04/10/2001    2   105     0   
3        M       1          257               0    0     0     0   
4        M       2          257      05/26/2009    7    48     0   

   DEPARTURE_HOUR ARRIVAL_HOUR  
0             9.1         13.5  
1             9.8         14.2  
2             9.8         14.2  
3            13.0         14.2  
4             8.2         15.8  

[5 rows x 35 columns]

In [27]:
flights.loc[flights.arr_gust == 'M', 'arr_gust'] = 0
flights.loc[flights.dept_gust == 'M', 'dept_gust'] = 0
flights.head()

MONTH  DOW  AIRLINE  ORIGIN_AIRPORT  DESTINATION_AIRPORT  DISTANCE  \
0      1    4        1              17                  277      1448   
1      1    4        1              17                  277      1448   
2      1    4        3              17                  277      1448   
3      1    4        1             121                  277       224   
4      1    4        0             176                  235      2330   

   ARRIVAL_DELAY  DIVERTED  CANCELLED dept_temp     ...      arr_vsby  \
0          -22.0         0          0      33.8     ...          10.0   
1          -14.0         0          0      33.8     ...          10.0   
2          -24.0         0          0      33.8     ...          10.0   
3            0.0         0          0      8.96     ...          10.0   
4           -9.0         0          0     46.04     ...         10.00   

  arr_gust arr_sky  arr_weather CertificateDate Manu Model  DAYS  \
0        0       1          257      12/18/2012    7    55     0   
1        0       1          257               0    0     0     0   
2        0       1          257      04/10/2001    2   105     0   
3        0       1          257               0    0     0     0   
4        0       2          257      05/26/2009    7    48     0   

   DEPARTURE_HOUR ARRIVAL_HOUR  
0             9.1         13.5  
1             9.8         14.2  
2             9.8         14.2  
3            13.0         14.2  
4             8.2         15.8  

[5 rows x 35 columns]

In [28]:
flights.columns

Index(['MONTH', 'DOW', 'AIRLINE', 'ORIGIN_AIRPORT', 'DESTINATION_AIRPORT',
       'DISTANCE', 'ARRIVAL_DELAY', 'DIVERTED', 'CANCELLED', 'dept_temp',
       'dept_dew', 'dept_hum', 'dept_windspd', 'dept_precip', 'dept_alti',
       'dept_vsby', 'dept_gust', 'dept_sky', 'dept_weather', 'arr_temp',
       'arr_dew', 'arr_hum', 'arr_windspd', 'arr_precip', 'arr_alti',
       'arr_vsby', 'arr_gust', 'arr_sky', 'arr_weather', 'CertificateDate',
       'Manu', 'Model', 'DAYS', 'DEPARTURE_HOUR', 'ARRIVAL_HOUR'],
      dtype='object')

In [ ]:
plane_birthdays = flights['CertificateDate'].tolist()
plane_age = []
for curr_day in plane_birthdays:
    day = str(curr_day)
    if day != '0' and day != 'None':
        #print(day)
        plane_age.append(2015 - int(day[-4:]))
    else:
        plane_age.append('-1')
print (plane_age)